nixos/modules/services/web-servers/caddy/default.nix

```diff
@@ -4,111 +4,159 @@ with lib;
 
 let
   cfg = config.services.caddy;
-  vhostToConfig = vhostName: vhostAttrs: ''
-    ${vhostName} ${builtins.concatStringsSep " " vhostAttrs.serverAliases} {
-      ${vhostAttrs.extraConfig}
-    }
-  '';
-  configFile = pkgs.writeText "Caddyfile" (builtins.concatStringsSep "\n"
-    ([ cfg.config ] ++ (mapAttrsToList vhostToConfig cfg.virtualHosts)));
-
-  formattedConfig = pkgs.runCommand "formattedCaddyFile" { } ''
-    ${cfg.package}/bin/caddy fmt ${configFile} > $out
-  '';
-
-  tlsConfig = {
-    apps.tls.automation.policies = [{
-      issuers = [{
-        inherit (cfg) ca email;
-        module = "acme";
-      }];
-    }];
-  };
 
-  adaptedConfig = pkgs.runCommand "caddy-config-adapted.json" { } ''
-    ${cfg.package}/bin/caddy adapt \
-      --config ${formattedConfig} --adapter ${cfg.adapter} > $out
-  '';
-  tlsJSON = pkgs.writeText "tls.json" (builtins.toJSON tlsConfig);
-
-  # merge the TLS config options we expose with the ones originating in the Caddyfile
-  configJSON =
-    if cfg.ca != null then
-      let tlsConfigMerge = ''
-        {"apps":
-          {"tls":
-            {"automation":
-              {"policies":
-                (if .[0].apps.tls.automation.policies == .[1]?.apps.tls.automation.policies
-                 then .[0].apps.tls.automation.policies
-                 else (.[0].apps.tls.automation.policies + .[1]?.apps.tls.automation.policies)
-                 end)
-              }
-            }
-          }
-        }'';
-      in
-      pkgs.runCommand "caddy-config.json" { } ''
-        ${pkgs.jq}/bin/jq -s '.[0] * ${tlsConfigMerge}' ${adaptedConfig} ${tlsJSON} > $out
+  virtualHosts = attrValues cfg.virtualHosts;
+  acmeVHosts = filter (hostOpts: hostOpts.useACMEHost != null) virtualHosts;
+
+  mkVHostConf = hostOpts:
+    let
+      sslCertDir = config.security.acme.certs.${hostOpts.useACMEHost}.directory;
+    in
       ''
-    else
-      adaptedConfig;
+        ${hostOpts.hostName} ${concatStringsSep " " hostOpts.serverAliases} {
+          bind ${concatStringsSep " " hostOpts.listenAddresses}
```

Comment:
@aanderse I noticed that by default `listenAddresses` is empty, which makes Caddy fail to bind. See also https://github.com/caddyserver/caddy/issues/4508

Reply:
> I noticed that by default `listenAddresses` is empty, which makes Caddy fail to bind.

I didn't notice this... I specifically read `caddy` documentation and tested this. When left blank it uses the default behavior. Verified again right now against `master` revision [___](https://github.com/NixOS/nixpkgs/commit/f6ee491278eb30e4619253fef211f1c1a92f8f65) using this `configuration.nix` file:

```nix
{ config, lib, pkgs, ... }:
let
  root = pkgs.writeTextDir "index.html" ''
    <html>
      <body>
        <p>it works</p>
      </body>
    </html>
  '';
in
{
  networking.firewall.allowedTCPPorts = [ 80 443 ];

  services.caddy.enable = true;
  services.caddy.virtualHosts.localhost = {
    extraConfig = ''
      root * ${root}
      file_server
    '';
  };
}
```

Reply:
Ohh it's actually because I was using an older nix expr of caddy that didn't include upstream systemd service. As a result I wasn't able to bind to :80 or :443. Ugh, if only overriding Go packages was straightforward, I wouldn't have use that nix expr in the first place.

Reply:
Glad to hear! I'll mark this as resolved. Thank you for confirming.